File: sdk/python/portals/transport.py
"""Length-framed transport for the Python worker side.

Matches the BEAM's `packet: 4` / `{:packet, 4}` framing exactly: each
frame is a 4-byte big-endian length prefix followed by that many bytes of
MessagePack-encoded envelope. `FramedSocket` hides that framing so callers
only ever see whole envelopes.
"""

import socket
import struct
import sys
from typing import Optional


class ConnectionClosed(Exception):
    pass


class FramedSocket:
    def __init__(self, sock: socket.socket):
        self._sock = sock
        self._buffer = b""
        self._lock_send = __import__("threading").Lock()

    @classmethod
    def connect_unix(cls, path: str) -> "FramedSocket":
        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        sock.connect(path)
        return cls(sock)

    def send_frame(self, payload: bytes) -> None:
        header = struct.pack(">I", len(payload))
        with self._lock_send:
            self._sock.sendall(header + payload)

    def recv_frame(self) -> Optional[bytes]:
        header = self._recv_exact(4)
        if header is None:
            return None
        (length,) = struct.unpack(">I", header)
        payload = self._recv_exact(length)
        if payload is None:
            raise ConnectionClosed("connection closed mid-frame")
        return payload

    def _recv_exact(self, n: int) -> Optional[bytes]:
        while len(self._buffer) < n:
            chunk = self._sock.recv(max(4096, n))
            if not chunk:
                if self._buffer:
                    raise ConnectionClosed("connection closed mid-frame")
                return None
            self._buffer += chunk

        data, self._buffer = self._buffer[:n], self._buffer[n:]
        return data

    def close(self) -> None:
        try:
            self._sock.close()
        except OSError:
            pass


class FramedStdio:
    """Experimental fallback transport: the same 4-byte length-prefix
    framing as `FramedSocket`, but carried over stdin/stdout instead of a
    Unix-domain socket, for platforms without Unix-socket support."""

    def __init__(self):
        self._stdin = sys.stdin.buffer
        self._stdout = sys.stdout.buffer
        self._lock_send = __import__("threading").Lock()

    def send_frame(self, payload: bytes) -> None:
        header = struct.pack(">I", len(payload))
        with self._lock_send:
            self._stdout.write(header + payload)
            self._stdout.flush()

    def recv_frame(self) -> Optional[bytes]:
        header = self._recv_exact(4)
        if header is None:
            return None
        (length,) = struct.unpack(">I", header)
        payload = self._recv_exact(length)
        if payload is None:
            raise ConnectionClosed("connection closed mid-frame")
        return payload

    def _recv_exact(self, n: int) -> Optional[bytes]:
        buf = b""
        while len(buf) < n:
            chunk = self._stdin.read(n - len(buf))
            if not chunk:
                if buf:
                    raise ConnectionClosed("connection closed mid-frame")
                return None
            buf += chunk
        return buf

    def close(self) -> None:
        pass


def socket_path_from_argv() -> str:
    if len(sys.argv) < 2:
        raise SystemExit("portals worker: expected the socket path as the last argument")
    return sys.argv[-1]

File: sdk/ruby/lib/portals/transport.rb
# frozen_string_literal: true

require 'socket'

# Length-framed transport for the Ruby worker side.
#
# Matches the BEAM's `{:packet, 4}` framing exactly: a 4-byte big-endian
# length prefix followed by that many bytes of MessagePack-encoded envelope.
# `FramedSocket` hides that framing so callers only ever see whole
# envelopes. `FramedStdio` is the fallback for platforms without Unix-socket
# support.
module Portals
  class ConnectionClosed < StandardError; end

  # Shared framing logic over any IO-ish object.
  module Framing
    def send_frame(payload)
      payload = payload.b
      @send_mutex.synchronize do
        write_all([payload.bytesize].pack('N') + payload)
      end
    end

    def recv_frame
      header = read_exact(4)
      return nil if header.nil?

      length = header.unpack1('N')
      payload = read_exact(length)
      raise ConnectionClosed, 'connection closed mid-frame' if payload.nil?

      payload
    end

    private

    def read_exact(count)
      buffer = +''.b
      while buffer.bytesize < count
        chunk = read_some(count - buffer.bytesize)
        if chunk.nil? || chunk.empty?
          raise ConnectionClosed, 'connection closed mid-frame' unless buffer.empty?

          return nil
        end
        buffer << chunk.b
      end
      buffer
    end
  end

  class FramedSocket
    include Framing

    def self.connect_unix(path)
      new(UNIXSocket.new(path))
    end

    def initialize(socket)
      @socket = socket
      @send_mutex = Mutex.new
    end

    def close
      @socket.close
    rescue IOError, SystemCallError
      nil
    end

    private

    def write_all(bytes) = @socket.write(bytes)

    def read_some(count)
      @socket.readpartial(count)
    rescue EOFError
      nil
    end
  end

  # Experimental fallback transport: identical framing carried over
  # stdin/stdout.
  class FramedStdio
    include Framing

    def initialize(input = $stdin, output = $stdout)
      @input = input
      @output = output
      @input.binmode
      @output.binmode
      @send_mutex = Mutex.new
    end

    def close = nil

    private

    def write_all(bytes)
      @output.write(bytes)
      @output.flush
    end

    def read_some(count) = @input.read(count)
  end

  module_function

  # `Portals.Connection` appends the socket path as the final CLI argument.
  def socket_path_from_argv
    raise 'portals worker: expected the socket path as the last argument' if ARGV.empty?

    ARGV.last
  end
end
